### ISSUE TITLE
Attachments not found with plugin v2.1.2

### ISSUE THREAD

**Issue by r66r** [10up/ElasticPress#693]
In a previous version (v1.8.1), the request below gave me 1000 results, included custom post types and attachaments. In the version v2.1.2, after reindexing, the same request give me only 4 results: only the custom post types are retrieved in the results and no attachments.

The WP_Query args:
```
array(6) {
  ["s"]=>
  string(8) "rectorat"
  ["post_type"]=>
  array(7) {
    [0]=>
    string(7) "archive"
    [1]=>
    string(5) "serie"
    [2]=>
    string(4) "book"
    [3]=>
    string(10) "periodical"
    [4]=>
    string(11) "documentary"
    [5]=>
    string(9) "inventory"
    [6]=>
    string(10) "attachment"
  }
  ["search_fields"]=>
  array(4) {
    [0]=>
    string(10) "post_title"
    [1]=>
    string(12) "post_content"
    ["taxonomies"]=>
    array(3) {
      [0]=>
      string(8) "typology"
      [1]=>
      string(14) "classification"
      [2]=>
      string(10) "collection"
    }
    ["meta"]=>
    array(7) {
      [0]=>
      string(5) "issue"
      [1]=>
      string(6) "author"
      [2]=>
      string(9) "publisher"
      [3]=>
      string(6) "paging"
      [4]=>
      string(8) "location"
      [5]=>
      string(10) "references"
      [6]=>
      string(10) "identifier"
    }
  }
  ["orderby"]=>
  string(9) "relevance"
  ["order"]=>
  string(4) "DESC"
  ["posts_per_page"]=>
  int(1000)
}
```

Note: I made an upgrade from v1.8.1 to v2.1.2, is there some issues I missed during the migration?

**Comment by r66r** [comment]
Note that my issue is not directly linked to issue #613.

Here's the filter I use to include attachment in indexing.

```
function uwa_core_elasticpress_set_post_statuses( $post_statuses ) {
	$post_statuses[] = 'inherit';
	return $post_statuses;
}
add_filter( 'ep_indexable_post_status', 'uwa_core_elasticpress_set_post_statuses' );
```

And after analyse, the attachments are correctly indexed... they are only not retrieved with the WP_Query request.

**Comment by r66r** [comment]
So, the first issue I found is that durind the search.php setup, my query vars are replaced with the default one.

At WP_Query first call:
```
WP_Query Object
(
    [query] => Array
        (
            [s] => rectorat
            [post_type] => Array
                (
                    [0] => archive
                    [1] => serie
                    [2] => book
                    [3] => periodical
                    [4] => documentary
                    [5] => inventory
                    [6] => attachment
                )

            [search_fields] => Array
                (
                    [0] => post_title
                    [1] => post_content
                    [taxonomies] => Array
                        (
                            [0] => typology
                            [1] => classification
                            [2] => collection
                        )

                    [meta] => Array
                        (
                            [0] => issue
                            [1] => author
                            [2] => publisher
                            [3] => paging
                            [4] => location
                            [5] => references
                            [6] => identifier
                        )

                )

            [orderby] => relevance
            [order] => DESC
            [posts_per_page] => 1000
        )

    [query_vars] => Array
        (
            [s] => rectorat
            [post_type] => Array
                (
                    [0] => archive
                    [1] => serie
                    [2] => book
                    [3] => periodical
                    [4] => documentary
                    [5] => inventory
                    [6] => attachment
                )

            [search_fields] => Array
                (
                    [0] => post_title
                    [1] => post_content
                    [taxonomies] => Array
                        (
                            [0] => typology
                            [1] => classification
                            [2] => collection
                        )

                    [meta] => Array
                        (
                            [0] => issue
                            [1] => author
                            [2] => publisher
                            [3] => paging
                            [4] => location
                            [5] => references
                            [6] => identifier
                        )

                )

            [orderby] => relevance
            [order] => DESC
            [posts_per_page] => 1000
            [error] => 
            [m] => 
            [p] => 0
            [post_parent] => 
            [subpost] => 
            [subpost_id] => 
            [attachment] => 
            [attachment_id] => 0
            [name] => 
            [static] => 
            [pagename] => 
            [page_id] => 0
            [second] => 
            [minute] => 
            [hour] => 
            [day] => 0
            [monthnum] => 0
            [year] => 0
            [w] => 0
            [category_name] => 
            [tag] => 
            [cat] => 
            [tag_id] => 
            [author] => 
            [author_name] => 
            [feed] => 
            [tb] => 
            [paged] => 0
            [meta_key] => 
            [meta_value] => 
            [preview] => 
            [sentence] => 
            [title] => 
            [fields] => 
            [menu_order] => 
            [embed] => 
            [category__in] => Array
                (
                )

            [category__not_in] => Array
                (
                )

            [category__and] => Array
                (
                )

            [post__in] => Array
                (
                )

            [post__not_in] => Array
                (
                )

            [post_name__in] => Array
                (
                )

            [tag__in] => Array
                (
                )

            [tag__not_in] => Array
                (
                )

            [tag__and] => Array
                (
                )

            [tag_slug__in] => Array
                (
                )

            [tag_slug__and] => Array
                (
                )

            [post_parent__in] => Array
                (
                )

            [post_parent__not_in] => Array
                (
                )

            [author__in] => Array
                (
                )

            [author__not_in] => Array
                (
                )

        )

    [tax_query] => WP_Tax_Query Object
        (
            [queries] => Array
                (
                )

            [relation] => AND
            [table_aliases:protected] => Array
                (
                )

            [queried_terms] => Array
                (
                )

            [primary_table] => 
            [primary_id_column] => 
        )

    [meta_query] => 
    [date_query] => 
    [post_count] => 0
    [current_post] => -1
    [in_the_loop] => 
    [comment_count] => 0
    [current_comment] => -1
    [found_posts] => 0
    [max_num_pages] => 0
    [max_num_comment_pages] => 0
    [is_single] => 
    [is_preview] => 
    [is_page] => 
    [is_archive] => 
    [is_date] => 
    [is_year] => 
    [is_month] => 
    [is_day] => 
    [is_time] => 
    [is_author] => 
    [is_category] => 
    [is_tag] => 
    [is_tax] => 
    [is_search] => 1
    [is_feed] => 
    [is_comment_feed] => 
    [is_trackback] => 
    [is_home] => 
    [is_404] => 
    [is_embed] => 
    [is_paged] => 
    [is_admin] => 
    [is_attachment] => 
    [is_singular] => 
    [is_robots] => 
    [is_posts_page] => 
    [is_post_type_archive] => 
    [query_vars_hash:WP_Query:private] => 69a53f74086345560250d70ae185b02b
    [query_vars_changed:WP_Query:private] => 
    [thumbnails_cached] => 
    [stopwords:WP_Query:private] => 
    [compat_fields:WP_Query:private] => Array
        (
            [0] => query_vars_hash
            [1] => query_vars_changed
        )

    [compat_methods:WP_Query:private] => Array
        (
            [0] => init_query_flags
            [1] => parse_tax_query
        )

)
```

After the 'add_action( 'pre_get_posts', 'ep_improve_default_search', 10, 1 );' action in search.php:
```
WP_Query Object
(
    [query] => Array
        (
            [s] => rectorat
            [post_type] => Array
                (
                    [0] => archive
                    [1] => serie
                    [2] => book
                    [3] => periodical
                    [4] => documentary
                    [5] => inventory
                    [6] => attachment
                )

            [search_fields] => Array
                (
                    [0] => post_title
                    [1] => post_content
                    [taxonomies] => Array
                        (
                            [0] => typology
                            [1] => classification
                            [2] => collection
                        )

                    [meta] => Array
                        (
                            [0] => issue
                            [1] => author
                            [2] => publisher
                            [3] => paging
                            [4] => location
                            [5] => references
                            [6] => identifier
                        )

                )

            [orderby] => relevance
            [order] => DESC
            [posts_per_page] => 1000
        )

    [query_vars] => Array
        (
            [s] => rectorat
            [post_type] => Array
                (
                    [0] => archive
                    [1] => serie
                    [2] => book
                    [3] => periodical
                    [4] => documentary
                    [5] => inventory
                    [6] => attachment
                )

            [search_fields] => Array
                (
                    [0] => post_title
                    [1] => post_content
                    [2] => post_excerpt
                    [3] => author_name
                    [taxonomies] => Array
                        (
                            [0] => post_tag
                            [1] => category
                        )

                )

            [orderby] => relevance
            [order] => DESC
            [posts_per_page] => 1000
            [error] => 
            [m] => 
            [p] => 0
            [post_parent] => 
            [subpost] => 
            [subpost_id] => 
            [attachment] => 
            [attachment_id] => 0
            [name] => 
            [static] => 
            [pagename] => 
            [page_id] => 0
            [second] => 
            [minute] => 
            [hour] => 
            [day] => 0
            [monthnum] => 0
            [year] => 0
            [w] => 0
            [category_name] => 
            [tag] => 
            [cat] => 
            [tag_id] => 
            [author] => 
            [author_name] => 
            [feed] => 
            [tb] => 
            [paged] => 0
            [meta_key] => 
            [meta_value] => 
            [preview] => 
            [sentence] => 
            [title] => 
            [fields] => 
            [menu_order] => 
            [embed] => 
            [category__in] => Array
                (
                )

            [category__not_in] => Array
                (
                )

            [category__and] => Array
                (
                )

            [post__in] => Array
                (
                )

            [post__not_in] => Array
                (
                )

            [post_name__in] => Array
                (
                )

            [tag__in] => Array
                (
                )

            [tag__not_in] => Array
                (
                )

            [tag__and] => Array
                (
                )

            [tag_slug__in] => Array
                (
                )

            [tag_slug__and] => Array
                (
                )

            [post_parent__in] => Array
                (
                )

            [post_parent__not_in] => Array
                (
                )

            [author__in] => Array
                (
                )

            [author__not_in] => Array
                (
                )

            [cache_results] => 
        )

    [tax_query] => WP_Tax_Query Object
        (
            [queries] => Array
                (
                )

            [relation] => AND
            [table_aliases:protected] => Array
                (
                )

            [queried_terms] => Array
                (
                )

            [primary_table] => 
            [primary_id_column] => 
        )

    [meta_query] => 
    [date_query] => 
    [post_count] => 0
    [current_post] => -1
    [in_the_loop] => 
    [comment_count] => 0
    [current_comment] => -1
    [found_posts] => 0
    [max_num_pages] => 0
    [max_num_comment_pages] => 0
    [is_single] => 
    [is_preview] => 
    [is_page] => 
    [is_archive] => 
    [is_date] => 
    [is_year] => 
    [is_month] => 
    [is_day] => 
    [is_time] => 
    [is_author] => 
    [is_category] => 
    [is_tag] => 
    [is_tax] => 
    [is_search] => 1
    [is_feed] => 
    [is_comment_feed] => 
    [is_trackback] => 
    [is_home] => 
    [is_404] => 
    [is_embed] => 
    [is_paged] => 
    [is_admin] => 
    [is_attachment] => 
    [is_singular] => 
    [is_robots] => 
    [is_posts_page] => 
    [is_post_type_archive] => 
    [query_vars_hash:WP_Query:private] => 69a53f74086345560250d70ae185b02b
    [query_vars_changed:WP_Query:private] => 
    [thumbnails_cached] => 
    [stopwords:WP_Query:private] => 
    [compat_fields:WP_Query:private] => Array
        (
            [0] => query_vars_hash
            [1] => query_vars_changed
        )

    [compat_methods:WP_Query:private] => Array
        (
            [0] => init_query_flags
            [1] => parse_tax_query
        )

)
```

**Comment by r66r** [comment]
This first issue is caused bay the following function, where `WP_Query->query_vars` are replaced without a check before.
```
function ep_improve_default_search( $query ) {
	if ( is_admin() ) {
		return;
	}

	/**
	 * Make sure this is an ElasticPress search query
	 */
	if ( ! ep_elasticpress_enabled( $query ) || ! $query->is_search() ) {
		return;
	}

	$query->set( 'search_fields', array(
		'post_title',
		'post_content',
		'post_excerpt',
		'author_name',
		'taxonomies' => array(
			'post_tag',
			'category',
		),
    ) );
}
```

So, during the format process in `EP_API->format_args()`, the requested fields are not correctly mapped for ElasticSearch.

**Comment by r66r** [comment]
The second issue is caused because I didn't add the `post_status` parameters for WP_Query. To get attachments, you have to add the `inherit` status in the query args.

**Comment by r66r** [comment]
Note that the comment in `EP_API->format_args()` isn't correct, because if no `post_status` is specified, the default is `publish`and not `any`.
```
/**
		 * Like WP_Query in search context, if no post_status is specified we default to "any". To
		 * be safe you should ALWAYS specify the post_status parameter UNLIKE with WP_Query.
		 *
		 * @since 2.1
		 */
		if ( ! empty( $args['post_status'] ) ) {
			// should NEVER be "any" but just in case
			if ( 'any' !== $args['post_status'] ) {
				$post_status = (array) $args['post_status'];
				$terms_map_name = 'terms';
				if ( count( $post_status ) < 2 ) {
					$terms_map_name = 'term';
					$post_status = $post_status[0];
 				}

				$filter['bool']['must'][] = array(
					$terms_map_name => array(
						'post_status' => $post_status,
					),
				);

				$use_filters = true;
			}
		} else {
			$filter['bool']['must'][] = array(
				'term' => array(
					'post_status' => 'publish',
				),
			);

			$use_filters = true;
		}
```

**Comment by r66r** [comment]
I think, what to do in a future release as enhancement, are:

1. Better check if the WP_Query is from the main query loop, or if it's a new WP_Query. In that case, don't replace que query_vars.

2. Change the default `post_status` to `any` if no `post_status` is set in the query vars.

**Comment by ivankristianto** [comment]
Hey @r66r 
thank you for your feedback.

For your first issue, i think it is fixed in EP 2.2, i saw it in this line of code: https://github.com/10up/ElasticPress/blob/develop/features/search/search.php#L88

For the second issue, I think it's not wise to set the `post_status` to `any` as default, but you can set it in the `post_status` in your custom WP_Query.
